### PR TITLE
GSOC 2014: Computational Photography 

### PR DESCRIPTION
**Pull request by Siddharthk**
a) Implementation of Intrinsic images. 
b) Implementation of Color transfer algorithms.
c) Implementation of Color constancy algorithms.

All modules added with test, documentation and demos.

### CODE REVIEW

File: modules/photo/src/color_transfer.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    Mat img1 = Mat(source.size(),CV_32FC3);
+    source.convertTo(img1,CV_32FC3,1.0/255.0);
+
+    Mat img2 = Mat(target.size(),CV_32FC3);
+    target.convertTo(img2,CV_32FC3,1.0/255.0);
+
+    if(flag == 0)
+    {
+        cvtColor(img1,img1,COLOR_BGR2Lab);
```
Comment (paroj) [inline modules/photo/src/color_transfer.cpp]: note that this is not the same as the L alpha beta color space used by reinhardt

File: modules/photo/src/color_transfer.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        cvtColor(img2,img2,COLOR_BGR2Lab);
+    }
+
+    int cols = source.cols;
+    Mat source_t = img1.t();
+    Mat target_t = img2.t();
+
+    cv::Mat src_reshaped = source_t.reshape ( 3, 1 );
```
Comment (paroj) [inline modules/photo/src/color_transfer.cpp]: why do you reshape?

File: modules/photo/src/color_transfer.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    cv::Mat trg_reshaped = target_t.reshape ( 3, 1 );
+
+    Scalar mean_source,stddev_source;
+    meanStdDev(src_reshaped,mean_source,stddev_source);
+
+    Scalar mean_target,stddev_target;
+    meanStdDev(trg_reshaped,mean_target,stddev_target);
+
+    vector <Mat> planes;
+    split(src_reshaped,planes);
```
Comment (paroj) [inline modules/photo/src/color_transfer.cpp]: splitting and merging is slow. better use Mat::forEach

**Comment by vpisarev** [comment]
the PR is too old; let's close it for now

**Comment by Siddharthk** [comment]
Let me know if any changes are required. I will work on that.

**Comment by vrabaud** [comment]
that was a great GSOC, any way we can merge this code ?

**Comment by Siddharthk** [comment]
Yes sure. Let me know how to proceed.

**Comment by mshabunin** [comment]
@Siddharthk , I've transferred part of your PR to the opencv_contrib repository: https://github.com/opencv/opencv_contrib/compare/master...mshabunin:transfer-3115
Feel free to adopt and continue working on it and create pull request.

There are several questions/things left:
- no weighted mode - is it possible to join both variants into one?
- no tests - is there a better way to test algorithms without comparing input/output images with reference (e.g. check some metric or other simple evaluation)?
- no samples for colorTransfer and intrinsicDecompose - they were too basic
- I have not checked algorithms flow - there can be some comments still

Notable changes:
- placed to the xphoto module with other white balance algorithms implemented this summer
- added support for new algorithms to color_balance_benchmark.py script and color_balance.cpp sample, I will run benchmark script in nearest time to compare with other algorithms
- documentation has been converted to doxygen
- used common WhiteBalancer interface
- added predefined modes for commonly used configuration variants

**Comment by mshabunin** [comment]
@Siddharthk , here is benchmark results:

Algorithm Name | Average Time | Mean error | Median error | 5th percentile | 20th percentile | 80th percentile | 95th percentile
--- | --- | --- | --- | --- | --- | --- | ---
grayworld | 6.83 ms | 4.74° | 3.51° | 0.54° | 1.48° | 7.41° | 12.79°
learning_based | 15.49 ms | 3.92° | 2.86° | 0.75° | 1.37° | 6.04° | 10.87°
grayedge:maxrgb | 154.24 ms | 10.66° | 10.85° | 2.64° | 6.35° | 15.17° | 16.99°
grayedge:shades | 149.30 ms | 4.19° | 2.98° | 0.38° | 1.14° | 7.10° | 11.16°
grayedge:grayedge | 330.36 ms | 6.30° | 6.04° | 1.24° | 2.93° | 9.34° | 12.59°
grayedge:general | 196.24 ms | 4.19° | 3.00° | 0.38° | 1.14° | 7.10° | 11.16°
grayedge:grayedge2 | 458.35 ms | 11.84° | 10.81° | 5.28° | 8.36° | 15.07° | 21.74°
grayedge:grayworld | 150.10 ms | 5.42° | 4.40° | 0.48° | 1.57° | 8.65° | 14.27°

**Pull request closed without merging**